File: Shared/Sources/KassaShared/Money.swift
import Foundation

/// Geldbetrag in ganzen Cent. Bewusst kein `Double` — Rundungsfehler haben in
/// einer Kassa nichts verloren.
public struct Money: Codable, Hashable, Sendable, Comparable {
    public var cents: Int

    public init(cents: Int) {
        self.cents = cents
    }

    public static let zero = Money(cents: 0)

    public static func < (lhs: Money, rhs: Money) -> Bool { lhs.cents < rhs.cents }
    public static func + (lhs: Money, rhs: Money) -> Money { Money(cents: lhs.cents + rhs.cents) }
    public static func - (lhs: Money, rhs: Money) -> Money { Money(cents: lhs.cents - rhs.cents) }
    public static func * (lhs: Money, rhs: Int) -> Money { Money(cents: lhs.cents * rhs) }

    public static func += (lhs: inout Money, rhs: Money) { lhs = lhs + rhs }

    /// Codiert als reine Cent-Zahl, damit über die Leitung nie ein Float geht.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.singleValueContainer()
        self.cents = try container.decode(Int.self)
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(cents)
    }
}

extension Money {
    /// Österreichische Schreibweise, z. B. `6,20 €`.
    public var formatted: String {
        let sign = cents < 0 ? "-" : ""
        let absolute = abs(cents)
        let euro = absolute / 100
        let rest = absolute % 100
        return "\(sign)\(euro),\(String(format: "%02d", rest)) €"
    }

    /// Ohne Währungszeichen, z. B. `6,20` — für Eingabefelder.
    public var formattedPlain: String {
        let sign = cents < 0 ? "-" : ""
        let absolute = abs(cents)
        return "\(sign)\(absolute / 100),\(String(format: "%02d", absolute % 100))"
    }

    /// Parst `6,20`, `6.20`, `6` oder `,50`. Gibt `nil` bei Unsinn zurück.
    public init?(parsing text: String) {
        let trimmed = text
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: "€", with: "")
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: ",", with: ".")
        guard !trimmed.isEmpty else { return nil }

        let negative = trimmed.hasPrefix("-")
        let body = negative ? String(trimmed.dropFirst()) : trimmed
        let parts = body.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count <= 2 else { return nil }

        let euroPart = parts[0].isEmpty ? "0" : String(parts[0])
        guard let euro = Int(euroPart), euro >= 0 else { return nil }

        var centPart = 0
        if parts.count == 2 {
            let raw = String(parts[1])
            guard raw.count <= 2, raw.allSatisfy(\.isNumber) else { return nil }
            let padded = raw.padding(toLength: 2, withPad: "0", startingAt: 0)
            guard let parsed = Int(padded) else { return nil }
            centPart = parsed
        }

        let total = euro * 100 + centPart
        self.init(cents: negative ? -total : total)
    }
}

File: Android/app/src/main/java/at/heuriger/kassa/wire/Money.kt
package at.heuriger.kassa.wire

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

/**
 * Geldbetrag in ganzen Cent. Bewusst kein `Double` — Rundungsfehler haben in
 * einer Kassa nichts verloren.
 *
 * Ueber die Leitung geht eine nackte Zahl, nie ein Objekt und nie ein Float.
 */
@Serializable(with = MoneySerializer::class)
data class Money(val cents: Int) : Comparable<Money> {

    override fun compareTo(other: Money): Int = cents.compareTo(other.cents)

    operator fun plus(other: Money): Money = Money(cents + other.cents)

    operator fun minus(other: Money): Money = Money(cents - other.cents)

    operator fun times(factor: Int): Money = Money(cents * factor)

    /** Oesterreichische Schreibweise, z. B. `6,20 €`. */
    val formatted: String
        get() = "$formattedPlain €"

    /** Ohne Waehrungszeichen, z. B. `6,20` — fuer Eingabefelder. */
    val formattedPlain: String
        get() {
            val sign = if (cents < 0) "-" else ""
            val absolute = kotlin.math.abs(cents)
            val rest = absolute % 100
            return "$sign${absolute / 100},${if (rest < 10) "0$rest" else "$rest"}"
        }

    companion object {
        val ZERO = Money(0)

        /** Parst `6,20`, `6.20`, `6` oder `,50`. Gibt `null` bei Unsinn zurueck. */
        fun parse(text: String): Money? {
            val trimmed = text.trim().replace("€", "").trim().replace(",", ".")
            if (trimmed.isEmpty()) return null

            val negative = trimmed.startsWith("-")
            val body = if (negative) trimmed.drop(1) else trimmed
            val parts = body.split(".")
            if (parts.size > 2) return null

            val euroPart = parts[0].ifEmpty { "0" }
            if (!euroPart.all { it in '0'..'9' }) return null
            // Long, nicht Int: Swifts Int ist 64 Bit, Kotlins 32. Ohne das liefern
            // schon acht Ziffern einen still umgeschlagenen, teils negativen Betrag —
            // in einer Kassa der schlechtestmoegliche Ausgang.
            val euro = euroPart.toLongOrNull() ?: return null

            var centPart = 0L
            if (parts.size == 2) {
                val raw = parts[1]
                if (raw.length > 2 || !raw.all { it in '0'..'9' }) return null
                centPart = raw.padEnd(2, '0').toLongOrNull() ?: return null
            }

            val total = euro * 100 + centPart
            if (total > Int.MAX_VALUE) return null
            return Money((if (negative) -total else total).toInt())
        }
    }
}

/** Money liegt als reine Cent-Zahl auf der Leitung — exakt wie Swifts `singleValueContainer`. */
object MoneySerializer : KSerializer<Money> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("at.heuriger.kassa.wire.Money", PrimitiveKind.INT)

    override fun serialize(encoder: Encoder, value: Money) = encoder.encodeInt(value.cents)

    override fun deserialize(decoder: Decoder): Money = Money(decoder.decodeInt())
}
